### PR TITLE
docs(api/dart): add support for generation and display

### LINKED ISSUE #1880: Dart API docs: add support for generation and display
Supersedes #1593.
### Task breakdown

**Package dependencies:**
- [ ] [**BLOCKING**] Update to new [harp](https://github.com/sintaxi/harp) package which depends on latest [terraform](https://github.com/sintaxi/terraform) release. This is needed so that dots (`.`) can be supported in site URIs; something that is essential for Dart given that library names contain dots. See https://github.com/sintaxi/harp/pull/526 (/cc @sintaxi).<br>Blocked awaiting the new release.

**Gulp tasks / build scripts:**
- [x] New tasks to run `pub` and `dartdoc` over (existing) `angular-dart` repo, peer to ng.io repo.
- [x] `build-api-docs` should also do Dart API docs.
- [x] Shredder should run over Dart sources and generate example fragments.
- [x] Create script to patch local `terraform` package, while we await new harp release, as a means of **unblocking**.

**Generator package:**
- [x] New `tools/dart-api-builder`; stand alone tool re-using part of the basis of a Dart API dgeni package.

**Harp/Jade site infrastructure:**
- [x] New `public/docs/_layout-dart-api.jade`.
- [x] Updates to other Harp/Jade template files.

**API List directive:**
- [ ] Create new or update existing `<api-list>` directive:
  - [x] Display only filter categories that are relevant to Dart.
  - [ ] Make library header a link to the library "index" page.

**Site data and Jade file generation:**
- [x] Main API page
  - [x] Generate `api-list.json`.
  - [x] Add warning that API is preliminary and that code might still appear as TS.
- [x] Generate data (`_data.json`) and Jade files for
  - [x] For each library (i.e., top-level folder).
    - [x] `_data.json`
    - [x] Library "index" Jade page.
    - [x] Each member Jade page.
  - [x] For library subfolders (e.g., containing class member details).
    - [x] `_data.json`
    - [x] Folder "index" page (it actually has the same name as the container -- e.g., class name).
    - [x] Each member page.
- [x] API Jade file page contents:
  - [x] Attributes like name and type
  - [x] Breadcrumbs
  - [ ] Page header section tabs.
  - [x] Core page content excerpt from dartdoc-generated page
    - [x] Exclude the "left" nav menu (full ng2 package content list) from the core page content.
  - [x] `{@example ...}` tags are replaced by their designated example code fragments.
- [x] Filter generated API pages (configurable via regExp in build file): exclude entries from
  - [x] Non-public, developer/internal libraries.
  - [x] Top-level catch all `angular2` library -- otherwise every entry appears twice.
- [x] Support developer & production modes for data (JSON) file generation
  - [x] Developer mode (pretty-printed).
  - [x] [NOT implemented] ~~Production mode (compact).~~

cc: @naomiblack @kwalrath @ferhatb @keertip

### PR DESCRIPTION
**Pull request by chalin**
Fixes #1880, which has a detailed task breakdown.
Supersedes #1593.

Note about the API subsite design: I had to resort to using Jade extends/includes; hence disabling Harp _layout/partials.

cc: @naomiblack @kwalrath @ferhatb @keertip

### CODE REVIEW

**Comment by chalin** [comment]
@wardbell @Foxandxss : ready for review.

**Comment by chalin** [comment]
I forgot to mention that to build the docs do: an install, patch terraforma, and then check-deploy:
```bash
./scripts/install.sh
./scripts/patch.sh
gulp check-deploy --dgeni-log=warn  # drop the flag if you want `info` level messages as usual
```

**Comment by kwalrath** [comment]
Taking a look... First impression: nice!

The second thing I noticed is the lack of a language dropdown at /docs/dart/latest/api/ (or any page under it). Would it be possible to add it? (And add some reasonable behavior in the likely case that there's no equivalent page at the dart -> ts/js version of the URL.)

If a dropdown isn't practical, we should put "Dart" in that space, so it's easier to tell that you're in the Dart version of the API docs.

**Comment by chalin** [comment]
I chose to leave out the dropdown because IMHO maintaining a complete dart <-> ts/js mapping might be tricky and not really worth the effort (we can open a separate issue as a "feature request" so that we can track discussion about it). Leaving the dropdown there with only Dart is a good idea. I'll look into it.

**Comment by chalin** [comment]
@filipesilva @Foxandxss @wardbell : I just rebased. Note the following change:
- Renamed `'harp-compile'` to `'check'` (as in `check-deploy` without the `deploy` :).
- Created new `'harp-compile'` that _only_ does a harp compile; i.e., doesn't depend on `'build-docs'` (where as `'check'` does).

This makes it easier to selectively rebuild only parts of the docs and then quickly get harp to start compiling.

**Comment by filipesilva** [comment]
@chalin that change lgtm

**Comment by chalin** [comment]
@kwalrath : I've created #1895 for the API page language-cross-referencing issue. I'd like to suggest that nothing be done in the scope of this PR. We can explore the range of alternative solutions via #1895 and a follow up PR. Thoughts?

**Comment by kwalrath** [comment]
LGTM.

**Comment by kwalrath** [comment]
@filipesilva @Foxandxss @wardbell @naomiblack: Unless you object, I plan to merge this after confirming that I can, indeed, generate the docs. 2 pm Pacific time is my goal.

**Comment by wardbell** [comment]
Please **no gulp task called _check_**  Check _what_? I don't approve.  

I do understand the impulse. It comes down to naming. I strongly prefer one of these alternatives (in my preferred order)

1. an option on `harp-compile` such as `harp-compile --no-build`
1. new  `build-compile` task that is `check-deploy` without the deploy.

File: gulpfile.js [hunk 1]
```diff
@@ -596,10 +643,35 @@ gulp.task('git-changed-examples', ['_shred-devguide-examples'], function(){
...
+  gutil.log('Note: issuing this command directly from the command line will show harp comiple warnings.');
+  return execPromise(cmd);
+});
+
+gulp.task('serve-www', [], function() {
+  // Serve generated site.
+  return execPromise('npm run live-server ./www');
+});
+
+gulp.task('check', ['build-docs'], function() {
```
Comment (wardbell) [inline gulpfile.js]: This could be `build-compile`

**Comment by wardbell** [comment]
Please also update the READM.md to explain these gulp tasks-and-options.

Otherwise, LGTM

**Comment by chalin** [comment]
@wardbell : named changed to `build-compile`.

More install / travis / gulpfile changes are on their way (to support Dart API docs building). I'll update the README then.

**Comment by wardbell** [comment]
LGTM. @kwalrath is free to merge :+1:

**Comment by chalin** [comment]
Design note (cc @petebacondarwin): while originally I started writing a dgeni package for Dart API processing, I ran into problems -- due to my lack of mastery of the dgeni framework --- and so, in the hopes of making some progress, I switched to building a stand alone transformer from the dgeni package parts. Maybe this can be revised later, or, I might just write a revised API page generator in Dart :).

**Pull request merged**